src/core/net/private/geoipdatabase.cpp

```diff
@@ -214,402 +168,330 @@ QString GeoIPDatabase::lookup(const QHostAddress &hostAddr) const
 #else
     Q_IPV6ADDR addr = hostAddr.toIPv6Address();
 #endif
-    const quint32 nodeCount = static_cast<quint32>(m_geoIPData->index.size());
-    Node node = m_geoIPData->index[0];
+
+    static const int nodeSize = m_recordSize / 4; // in bytes
+    static const int indexSize = m_nodeCount * nodeSize;
+    static const int recordBytes = nodeSize / 2;
+    const uchar *ptr = m_data;
+
     for (int i = 0; i < 16; ++i) {
         for (int j = 0; j < 8; ++j) {
             bool right = static_cast<bool>((addr[i] >> (7 - j)) & 1);
-            quint32 id = (right ? node.right : node.left);
-            if (id == nodeCount)
+            // Interpret the left/right record as number
+            if (right)
+                ptr += recordBytes;
+
+            uchar buf[4] = { 0 };
+            memcpy(&buf[4 - recordBytes], ptr, recordBytes);
+            fromBigEndian(buf, 4);
+            quint32 id = *(reinterpret_cast<quint32 *>(buf));
+
+            if (id == m_nodeCount) {
                 return QString();
-            else if (id > nodeCount)
-                return m_geoIPData->countries[id];
-            else
-                node = m_geoIPData->index[id];
+            }
+            else if (id > m_nodeCount) {
+                QString country = m_countries.value(id);
+                if (country.isEmpty()) {
+                    const quint32 offset = id - m_nodeCount - sizeof(DATA_SECTION_SEPARATOR);
+                    quint32 tmp = offset + indexSize + sizeof(DATA_SECTION_SEPARATOR);
+                    QVariant val = readDataField(tmp);
+                    if (val.userType() == QMetaType::QVariantHash) {
+                        country = val.toHash()["country"].toHash()["iso_code"].toString();
+                        m_countries[id] = country;
```

Comment:
You made me remember. This is what prevents us from using const.
So I'll make it mutable as per your suggestion.